wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,310 @@
+    // The host of the site to which the authenticator is associated.
+    // This can be used to prevent credentials from being sent to other sites.
+    fn host(&self) -> &str;
+
+    async fn authentication_headers(&self) -> Option<HeaderMap>;
+
+    fn reset(&self);
+
+    fn should_authenticate(&self, request_url: &str, response_status_code: Option<u16>) -> bool;
+
+    async fn re_authenticate(
+        &self,
+        request: &WpNetworkRequest,
+        previous_response: &WpNetworkResponse,
+    ) -> Option<HeaderMap> {
+        if self.should_authenticate(&request.url.0, Some(previous_response.status_code)) {
+            self.reset();
+            return self.authentication_headers().await;
+        }
+
+        None
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct NilAuthenticator {}
+
+#[async_trait::async_trait]
+impl Authenticator for NilAuthenticator {
+    fn should_authenticate(&self, _request_url: &str, _response_status_code: Option<u16>) -> bool {
+        false
+    }
+
+    fn host(&self) -> &str {
+        ""
+    }
+
+    async fn authentication_headers(&self) -> Option<HeaderMap> {
+        None
+    }
+
+    fn reset(&self) {
+        // Do nothing.
+    }
+}
+
+#[derive(Debug)]
+pub struct ApplicationPasswordAuthenticator {
+    host: String,
+    token: String,
+}
+
+impl ApplicationPasswordAuthenticator {
+    pub fn new(host: String, token: String) -> Self {
+        Self { host, token }
+    }
+
+    pub fn with_application_password(host: String, username: String, password: String) -> Self {
+        use base64::prelude::*;
+        let token = BASE64_STANDARD.encode(format!("{}:{}", username, password));
+        Self::new(host, token)
+    }
+}
+
+#[async_trait::async_trait]
+impl Authenticator for ApplicationPasswordAuthenticator {
+    fn host(&self) -> &str {
+        self.host.as_str()
+    }
+
+    async fn authentication_headers(&self) -> Option<HeaderMap> {
+        let mut headers = HeaderMap::new();
+        headers.append(
+            http::header::AUTHORIZATION,
+            format!("Basic {}", self.token)
+                .try_into()
+                .expect("token is always a valid header value"),
+        );
+
+        Some(headers)
+    }
+
+    fn reset(&self) {
+        // Do nothing.
+    }
+
+    fn should_authenticate(&self, request_url: &str, response_status_code: Option<u16>) -> bool {
+        let is_same_host = Url::parse(request_url)
+            .ok()
+            .map(|f| f.host_str() == Some(self.host()));
+        if is_same_host != Some(true) {
+            return false;
+        }
+
+        // Authenticated has already been sent. No need to make a second attempt.
+        response_status_code.is_none()
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct CookieAuthenticator {
+    api_base_url: ApiBaseUrl,
+    credentials: WpLoginCredentials,
+    request_executor: std::sync::Arc<dyn RequestExecutor>,
+    nonce: RwLock<Option<String>>,
```

Comment:
Could you explain why we need/want to use `RwLock`?

I've gone through the implementation and it looks like we are switching from Rust's compile time borrow checker guarantees to the runtime checks with the `RwLock` and then immediately give up on them and crash any time there are more than 1 threads writing at the same time, or if there is a write & read happening at the same time.

Reply:
When there are concurrent requests, the nonce will be read/write concurrently. I think we'll have to use some sort of lock to prevent concurrency issues.

In https://github.com/Automattic/wordpress-rs/pull/327/commits/527c176fcc98dab5e1a20bcf8eee71544a967d6a, I have changed this `std::sync::RwLock` to `futures::lock::Mutex` to work with async functions better and prevent concurrent read & write to nonce.